feature/s3/transfermanager/api_op_GetObject.go

```diff
@@ -0,0 +1,927 @@
+	//
+	// This functionality is not supported for directory buckets.
+	//
+	// [Server-Side Encryption (Using Customer-Provided Encryption Keys)]: https://docs.aws.amazon.com/AmazonS3/latest/dev/ServerSideEncryptionCustomerKeys.html
+	SSECustomerKeyMD5 string
+
+	// Version ID used to reference a specific version of the object.
+	//
+	// By default, the GetObject operation returns the current version of an object.
+	// To return a different version, use the versionId subresource.
+	//
+	//   - If you include a versionId in your request header, you must have the
+	//   s3:GetObjectVersion permission to access a specific version of an object. The
+	//   s3:GetObject permission is not required in this scenario.
+	//
+	//   - If you request the current version of an object without a specific versionId
+	//   in the request header, only the s3:GetObject permission is required. The
+	//   s3:GetObjectVersion permission is not required in this scenario.
+	//
+	//   - Directory buckets - S3 Versioning isn't enabled and supported for directory
+	//   buckets. For this API operation, only the null value of the version ID is
+	//   supported by directory buckets. You can only specify null to the versionId
+	//   query parameter in the request.
+	//
+	// For more information about versioning, see [PutBucketVersioning].
+	//
+	// [PutBucketVersioning]: https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutBucketVersioning.html
+	VersionID string
+}
+
+func (i GetObjectInput) mapGetObjectInput(enableChecksumValidation bool) *s3.GetObjectInput {
+	input := &s3.GetObjectInput{
+		Bucket: aws.String(i.Bucket),
+		Key:    aws.String(i.Key),
+	}
+
+	if i.ChecksumMode != "" {
+		input.ChecksumMode = s3types.ChecksumMode(i.ChecksumMode)
+	} else if enableChecksumValidation {
+		input.ChecksumMode = s3types.ChecksumModeEnabled
+	}
+
+	if i.RequestPayer != "" {
+		input.RequestPayer = s3types.RequestPayer(i.RequestPayer)
+	}
+
+	input.ExpectedBucketOwner = nzstring(i.ExpectedBucketOwner)
+	input.IfMatch = nzstring(i.IfMatch)
+	input.IfNoneMatch = nzstring(i.IfNoneMatch)
+	input.IfModifiedSince = nztime(i.IfModifiedSince)
+	input.IfUnmodifiedSince = nztime(i.IfUnmodifiedSince)
+	input.ResponseCacheControl = nzstring(i.ResponseCacheControl)
+	input.ResponseContentDisposition = nzstring(i.ResponseContentDisposition)
+	input.ResponseContentEncoding = nzstring(i.ResponseContentEncoding)
+	input.ResponseContentLanguage = nzstring(i.ResponseContentLanguage)
+	input.ResponseContentType = nzstring(i.ResponseContentType)
+	input.ResponseExpires = nztime(i.ResponseExpires)
+	input.SSECustomerAlgorithm = nzstring(i.SSECustomerAlgorithm)
+	input.SSECustomerKey = nzstring(i.SSECustomerKey)
+	input.SSECustomerKeyMD5 = nzstring(i.SSECustomerKeyMD5)
+	input.VersionId = nzstring(i.VersionID)
+
+	return input
+}
+
+// GetObjectOutput represents a response from GetObject() or DownloadObject() call. It contains common fields
+// of s3 GetObject output
+type GetObjectOutput struct {
```

Comment:
We can't share these types between get and download. There's already divergence between what each one needs - as written, you have `Body` in the output for DownloadObject, which is functionally useless.

Reply:
(the same is true for the input)

Reply:
Yea, we need `io.WriterAt` in `DownloadObjectInput` as dst, but not for `GetObject`